faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,207 @@
+     * encode(Arrays.asList(1, 2)) => new ArrayV(Arrays.asList(1, 2))
+     * encode(new byte[] {1, 2}) => new ByteV(new byte[] {1, 2})
+     * encode(user) => new ObjectV(ImmutableMap.of("user_name", new StringV("john"), "password", new StringV("s3cr3t")))
+     * }</pre>
+     *
+     * @param obj Any instance of user defined classes, primitive values or any
+     *            generic collection like {@link java.util.List}, {@link java.util.Set} or {@link java.util.Map}
+     * @return A FaunaDB {@link Value} corresponding to the given argument
+     */
+    public static Value encode(Object obj) {
+        if (obj == null)
+            return Value.NullV.NULL;
+
+        if (Value.class.isInstance(obj))
+            return (Value) obj;
+
+        if (stack.contains(obj))
+            throw new FaunaException(format("Self reference loop detected for object `%s`", obj));
+
+        try {
+            stack.push(obj);
+
+            return encodeIntern(obj);
+        } finally {
+            stack.pop();
+        }
+    }
+
+    private static Value encodeIntern(Object obj) {
+        Class<?> clazz = obj.getClass();
+
+        if (clazz.isEnum())
+            return wrapEnum((Enum<?>) obj, clazz);
+
+        if (clazz == String.class)
+            return new StringV((String) obj);
+
+        if (clazz == Boolean.class)
+            return BooleanV.valueOf((Boolean) obj);
+
+        if (clazz == Character.class)
+            return new LongV((Character) obj);
+
+        if (clazz == Byte.class)
+            return new LongV((Byte) obj);
+
+        if (clazz == Short.class)
+            return new LongV((Short) obj);
+
+        if (clazz == Integer.class)
+            return new LongV((Integer) obj);
+
+        if (clazz == Long.class)
+            return new LongV((Long) obj);
+
+        if (clazz == Float.class)
+            return new DoubleV((Float) obj);
+
+        if (clazz == Double.class)
+            return new DoubleV((Double) obj);
+
+        if (clazz == LocalDate.class)
+            return new DateV((LocalDate) obj);
+
+        if (clazz == HighPrecisionTime.class)
+            return new TimeV((HighPrecisionTime) obj);
+
+        if (clazz == Instant.class)
+            return new TimeV(HighPrecisionTime.fromInstant((Instant) obj));
+
+        if (clazz == byte[].class)
+            return new BytesV((byte[]) obj);
+
+        if (clazz.isArray())
+            return wrapArray(obj);
+
+        if (Iterable.class.isAssignableFrom(clazz))
+            return wrapIterable((Iterable<?>) obj);
+
+        if (Map.class.isAssignableFrom(clazz))
+            return wrapMap((Map<?, ?>) obj);
+
+        return wrapObject(obj);
+    }
+
+    private static Value wrapEnum(Enum<?> obj, Class<?> enumClass) {
+        Map<Enum<?>, StringV> enumMap = ENUMS.get(enumClass);
+
+        if (enumMap == null) {
+            enumMap = newHashMap();
+
+            for (Enum en : (Enum[]) enumClass.getEnumConstants()) {
```

Comment:
Maybe we should wrap enums in a different class so we keep a single map. It'd be less duplicated.